src/printer/standard.rs

```diff
@@ -0,0 +1,970 @@
+            stats,
+            needs_match_granularity,
+        }
+    }
+
+    fn needs_match_granularity(&self) -> bool {
+        let supports_color = self.wtr.borrow().supports_color();
+        let match_colored = !self.config.colors.matched().is_none();
+
+        (supports_color && match_colored)
+            || self.config.column
+            || self.config.per_match
+            || self.config.only_matching
+            || self.config.stats
+    }
+}
+
+impl<W> Standard<W> {
+    pub fn has_written(&self) -> bool {
+        self.wtr.borrow().total_count() > 0
+    }
+
+    pub fn get_mut(&mut self) -> &mut W {
+        self.wtr.get_mut().get_mut()
+    }
+
+    pub fn into_inner(self) -> W {
+        self.wtr.into_inner().into_inner()
+    }
+}
+
+#[derive(Debug)]
+pub struct StandardSink<'p, 's, W> {
+    standard: &'s mut Standard<W>,
+    path: Option<PrinterPath<'p>>,
+    start_time: Instant,
+    match_count: u64,
+    after_context_remaining: u64,
+    stats: Option<Stats>,
+    needs_match_granularity: bool,
+}
+
+impl<'p, 's, W: WriteColor> StandardSink<'p, 's, W> {
+    pub fn has_match(&self) -> bool {
+        self.match_count > 0
+    }
+
+    pub fn match_count(&self) -> u64 {
+        self.match_count
+    }
+
+    pub fn stats(&self) -> Option<&Stats> {
+        self.stats.as_ref()
+    }
+
+    fn should_quit(&self) -> bool {
+        let limit = match self.standard.config.max_matches {
+            None => return false,
+            Some(limit) => limit,
+        };
+        if self.match_count < limit {
+            return false;
+        }
+        self.after_context_remaining == 0
+    }
+
+    fn match_more_than_limit(&self) -> bool {
+        let limit = match self.standard.config.max_matches {
+            None => return false,
+            Some(limit) => limit,
+        };
+        self.match_count > limit
+    }
+}
+
+impl<'p, 's, W: WriteColor> Sink for StandardSink<'p, 's, W> {
+    type Error = io::Error;
+
+    fn matched(&mut self, searcher: &Searcher, mat: &SinkMatch<'_>) -> Result<bool, io::Error> {
+        self.match_count += 1;
+        if self.match_more_than_limit() {
+            self.after_context_remaining = self.after_context_remaining.saturating_sub(1);
+        } else {
+            self.after_context_remaining = searcher.after_context() as u64;
+        }
+
+        if let Some(ref mut stats) = self.stats {
+            stats.add_matches(self.standard.matches.len() as u64);
+            stats.add_matched_lines(mat.lines().count() as u64);
+        }
+
+        StandardImpl::from_match(searcher, self, mat).sink()?;
+        Ok(!self.should_quit())
+    }
+
+    fn context(&mut self, searcher: &Searcher, ctx: &SinkContext<'_>) -> Result<bool, io::Error> {
+        self.standard.matches.clear();
+
+        if ctx.kind() == &SinkContextKind::After {
+            self.after_context_remaining = self.after_context_remaining.saturating_sub(1);
+        }
+        // if searcher.invert_match() {
```

Comment:
I left this commented out as sort of a "reminder" that I'm pretty sure that if we try and expose an invert-match option it won't "just work" (this is where in that case it was doing the weird "re-matching" to get exact match locations, which now the idea is they are getting passed explicitly to the sink at the time of telling it about a "match" (which may in practice be a "clump of match lines")